programs/native/template_signer/src/lib.rs

```diff
@@ -0,0 +1,579 @@
+//! TemplateSigner is a protocol to coordinate signing of a message that fits a specific template.
+//! Terminology
+//!     * `template` - The partially filled message that is to be signed by the signer.  The
+//!       template contains the Key for the kind of cryptographic key the message is signed with,
+//!       a mask that identifies the undefined regions of the message and signature to proove that
+//!       the signer has control of the Key.
+//!
+//!     * `signer` - The Pubkey of the entity that sign the template.  This is a Solana Pubkey and
+//!       this key is different and unreleated to the template.
+//!
+//!     * `escrow` - The tokens held by the SigningContract.
+//!
+//!     * `SigningContract` - An instance of the state machine for the specific template to be signed.
+//!
+//!     * `guarantee` - The tokens held as a guarantee that the `signer` will actually sign the
+//!       message when the request is made.
+//!
+//!     * `owner` - The current owner of the SigningContract.  Only the owner can make the request to
+//!     sign.
+//!
+//! A template is composed of an arbitratry key, a message, and a mask as well as a signature and a
+//! challenge to prove ownership of the key.
```

Comment:
nit: this is the first mention of “challenge” in the text, some reverence to how the “challenge” relates to the above definitions would help readability I think